File: lib/logic/hitung.dart
double hitungAnuitasBiasa(double M, double i, int N) {
  if (N == 0) return M;
  if (i == 0) return M / N;

  double pembilang = i;
  double penyebut = (1 - (1 + i) * -N);

  if (penyebut == 0) {
    return double.infinity;
  }
  return (pembilang / penyebut) * M;
}

double hitungAnuitasAwalPeriode(double M, double i, int N) {
  if (N == 0) return M;
  if (i == 0) return M / N;

  double pembilang = i;
  double penyebut = (1 - (1 + i) * -N);

  if (penyebut == 0) {
    return double.infinity;
  }
  return (pembilang / penyebut) * M * (1 + i);
}

double hitungAngsuranPokok(double A, double i, double sisaPinjamanSebelumnya) {
  double bunga = sisaPinjamanSebelumnya * i;
  return A - bunga;
}

double hitungBungaAngsuran(double sisaPinjamanSebelumnya, double i) {
  return sisaPinjamanSebelumnya * i;
}

double hitungSisaPinjamanSetelahAngsuranKeK(
    double M, double i, int N, int k, String tipeAnuitas) {
  double A;
  if (tipeAnuitas == "biasa") {
    A = hitungAnuitasBiasa(M, i, N);
  } else {
    A = hitungAnuitasAwalPeriode(M, i, N);
  }

  if (A.isInfinite) {
    return M;
  }

  double sisaPinjaman = M;
  for (int periode = 1; periode <= k; periode++) {
    double bunga = sisaPinjaman * i;
    double pokok = A - bunga;

    if (pokok < 0 && pokok.abs() < 1e-9) {
      pokok = 0;
    }
    if (pokok < 0) {
      return 0;
    }

    sisaPinjaman -= pokok;

    if (sisaPinjaman < 0 && sisaPinjaman.abs() < 1e-9) {
      sisaPinjaman = 0;
    }
    if (sisaPinjaman < 0) {
      return 0;
    }
  }
  return sisaPinjaman;
}

File: asalapp/logic/hitung.py
def hitung_anuitas_biasa(M, i, N):
    if (1 - (1 + i) ** -N) == 0: 
        return M / N if N != 0 else float('inf') 
    A = (i / (1 - (1 + i) ** -N)) * M
    return A

def hitung_anuitas_awal_periode(M, i, N):
    if (1 - (1 + i) ** -N) == 0: 
        return (M / N if N != 0 else float('inf'))
    A = (i / (1 - (1 + i) ** -N)) * M * (1 + i) 
    return A

def hitung_angsuran_pokok(A, i, sisa_pinjaman_sebelumnya):
    bunga = sisa_pinjaman_sebelumnya * i
    pokok = A - bunga
    return pokok

def hitung_bunga_angsuran(sisa_pinjaman_sebelumnya, i):
    return sisa_pinjaman_sebelumnya * i

def hitung_sisa_pinjaman_setelah_angsuran_ke_k(M, i, N, k, tipe_anuitas="biasa"):
    from .hitung import hitung_anuitas_biasa, hitung_anuitas_awal_periode
    if tipe_anuitas == "biasa":
        A = hitung_anuitas_biasa(M, i, N)
    else:
        A = hitung_anuitas_awal_periode(M, i, N)

    if A == float('inf'):
        return M

    sisa_pinjaman = M
    for _ in range(1, k + 1):
        bunga = sisa_pinjaman * i
        pokok = A - bunga
        if pokok < 0 and abs(pokok) < 1e-9:
            pokok = 0
        if pokok < 0:
            return 0
        sisa_pinjaman -= pokok
        if sisa_pinjaman < 0 and abs(sisa_pinjaman) < 1e-9:
            sisa_pinjaman = 0
        if sisa_pinjaman < 0:
            return 0
    return sisa_pinjaman